### ISSUE TITLE
Json formatter bypasses redirect_stdout

### ISSUE THREAD

**Issue by martmists** [PyCQA/pylint#3227]
### Steps to reproduce
1. Install pylint
2. Run it like this:

```py
from contextlib import redirect_stdout
from io import StringIO
from pylint.lint import Run

file = StringIO()
with redirect_stdout(file):
    Run(["some_file.py"], do_exit=False)  # No output


file_2 = StringIO()
with redirect_stdout(file_2):
    Run(["-f", "json", "some_file.py"], do_exit=False)  # JSON output
```

### Current behavior

There is still being written to stdout (I checked stderr with the same method, but according to the source stdout is being used)

### Expected behavior

`redirect_stdout` catches all content written to stdout and writes it to the StringIO instead

### pylint --version output

```
$ pylint --version
pylint 2.4.3
astroid 2.3.1
Python 3.7.4 (default, Oct  4 2019, 06:57:26) 
[GCC 9.2.0]
```

### Cause

[json_reporter.py:25](https://github.com/PyCQA/pylint/blob/master/pylint/reporters/json_reporter.py#L25)
JSONReporter has sys.stdout as default parameter, which causes it to keep a reference to the original output.
To solve this, simply replace sys.stdout with `None` as it will fetch sys.stdout if the output parameter is falsy.
